examples/src/bin/widget_subclass.rs

```diff
@@ -0,0 +1,109 @@
+use std::cell::RefCell;
+use std::env;
+
+use gio::prelude::*;
+use glib::glib_wrapper;
+use gtk::prelude::*;
+
+mod imp {
+    use super::*;
+    use glib::{glib_object_subclass, WeakRef};
+    use gtk::subclass::prelude::*;
+
+    #[derive(Debug)]
+    pub struct Button {
+        child: RefCell<WeakRef<gtk::Widget>>,
+    }
+
+    impl Drop for Button {
+        fn drop(&mut self) {
+            // Child widgets need to be manually unparented.
```

Comment:
I think this has to happen from `dispose()` and not `finalize()` actually, which then needs someone to think about how to do that in a safe way.

Reply:
Reason being `gtk_container_destroy()` etc calling `g_object_run_dispose()` and that supposedly cleaning up the whole widget hierarchy. IMHO a hack for not using weak references correctly in GTK but that's just how it is.

Reply:
Yeah, I was asking about this yesterday but couldn't find a way to hook into dispose.

Reply:
Needs changes in `glib/src/subclass` and figuring out a way how to safely expose that. At the moment `dispose()` is called the object is not fully existing anymore. It's basically a zombie. Calling any API on it can cause all kinds of interesting effects.

The way how GTK does the `g_object_run_dispose()` is unsafe by design and will probably cause some more headaches in Rust. Something to figure out.